libmariadb/mariadb_lib.c

```diff
@@ -2653,26 +2657,7 @@ mysql_optionsv(MYSQL *mysql,enum mysql_option option, ...)
     OPT_SET_VALUE_STR(&mysql->options, my_cnf_file, (char *)arg1);
     break;
   case MYSQL_READ_DEFAULT_GROUP:
-    if (!arg1 || !((char *)arg1)[0])
-    {
-#if defined(HAVE_PROGRAM_INVOCATION_SHORT_NAME)
-      const char * appname = program_invocation_short_name;
-#elif defined(HAVE_GETPROGNAME)
-      const char * appname = getprogname();
-#elif defined(WIN32)
-      char module_filename[MAX_PATH];
-      char appname[MAX_PATH]="";
-      if (GetModuleFileName(NULL, module_filename, MAX_PATH))
-      {
-        _splitpath(module_filename,NULL, NULL, appname, NULL);
-      }
-#else
-      const char * appname = "";
-#endif
-      OPT_SET_VALUE_STR(&mysql->options, my_cnf_group, appname);
-      break;
-    }
-    OPT_SET_VALUE_STR(&mysql->options, my_cnf_group, (char *)arg1);
+    OPT_SET_VALUE_STR(&mysql->options, my_cnf_group, arg1 ? (char *)arg1 : "");
```

Comment:
Why? I think it was an important feature, contributing to the standard expected behavior of all MariaDB clients

Reply:
- program name detection failed on several platforms
- In libmysql the default configuration groups will be read if you specify NULL or ""(empty string), but not the programname group.

Reply:
Well, this is exactly the point. the "standard expected behavior" is to read the default groups and the programname group. Now there's no way to do it. If a client specifies "" or NULL, C/C won't read the programname group. If the client specifies "programname" explicitly, C/C won't read default groups. So no way to get the standard behavior.

On what platforms did the program name detection fail?

Reply:
The current behavior of Connector/C and libmysql is that configuration files will be read if either config_file or config_group was specified:
```
/* use default options */
  if (mysql->options.my_cnf_file || mysql->options.my_cnf_group)
```
So the initial behavior for mysql_options(mysql, MYSQL_OPT_READ_DEFAULT_GROUP, "") was reading default configuration files with default groups, but not the program name.

Please note that also the my.cnf parser is very basic and will fail, e.g. if your application name has special chars, e.g. "]client"

Reply:
If I remember correctly, program name wasn't detected on Solaris (this was once fixed in master branch), but also when crosscompiling for android and arm.